### PR TITLE
Feature/zk proof flow

### PR DESCRIPTION
**Pull request by lukachi**

### CODE REVIEW

File: packages/auth-zkp-iden3/src/auth-zkp.ts [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    const [wasm, provingKey] = await Promise.all([
+      readBytesFile(AuthZkp.config.CIRCUIT_WASM_URL),
+      readBytesFile(AuthZkp.config.CIRCUIT_FINAL_KEY_URL),
+    ])
+
+    const jwzTokenRaw = await token2.prove(provingKey, wasm)
+
+    const { rawData: issuerData } = await api
+      .withBaseUrl(offerData.body.url)
+      .post<VerifiableCredentials<T>>('', {
```
Comment (napalmpapalam) [inline packages/auth-zkp-iden3/src/auth-zkp.ts:77]: why u don't use as generic `VerifiableCredentials<T>` here?

File: packages/auth-zkp-iden3/src/types/auth-zkp-types.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+      credentialStatus: {
+        id: string
+        revocationNonce: number
+        type: string
+      }
+      credentialSubject: {
+        id: string
+        type: string
+        // isMember: number
+        // serialNumber: number
```
Comment (napalmpapalam) [inline packages/auth-zkp-iden3/src/types/auth-zkp-types.ts:45]: /

File: packages/auth-zkp-iden3/src/types/auth-zkp-types.ts [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  }
+  from: string
+  id: string
+  thid: string
+  to: string
+  typ: string
+  type: string
+}
+
+export type AuthZkpConfig = {
```
Comment (napalmpapalam) [inline packages/auth-zkp-iden3/src/types/auth-zkp-types.ts:96]: just `Config`

File: packages/identity-gen-iden3/README.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
+# @rarimo/identity-gen-iden3
+These packages aim to provide developers with a set of commonly used functions and features for building web applications, such as handling big numbers, date manipulation, subscribing to and receiving notifications when certain events occur with EventEmitter, and more.
```
Comment (napalmpapalam) [inline packages/identity-gen-iden3/README.md:2]: /

File: packages/identity-gen-iden3/README.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  { find: 'ethers', replacement: '../../node_modules/ethers/dist/ethers.esm.js' },
+  { find: 'util', replacement: '../../node_modules/util/util.js' },
+  { find: 'ejc', replacement: '../../node_modules/ejs/ejs.min.js' },
+  { find: 'snarkjs', replacement: '../../node_modules/snarkjs/build/snarkjs.min.js' },
+  { find: "@iden3/js-iden3-core", replacement: "../../node_modules/@iden3/js-iden3-core/dist/esm_esbuild/index.js" },
+  { find: "@iden3/js-jwz", replacement: "../../node_modules/@iden3/js-jwz/dist/esm_esbuild/index.js" },
+  { find: "@iden3/js-crypto", replacement: "../../node_modules/@iden3/js-crypto/dist/esm_esbuild/index.js" },
+  { find: "@iden3/js-jsonld-merklization", replacement: "../../node_modules/@iden3/js-jsonld-merklization/dist/esm_esbuild/index.js" }
+]
+```
```
Comment (napalmpapalam) [inline packages/identity-gen-iden3/README.md:39]: move this to the main readme

File: packages/identity-gen-iden3/src/helpers/identity-helpers.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
+import { Hex, PrivateKey } from '@iden3/js-crypto'
+
+export const initPrivateKey = (hexString?: string): string => {
```
Comment (napalmpapalam) [inline packages/identity-gen-iden3/src/helpers/identity-helpers.ts:3]: `parsePrivateKey`

File: packages/identity-gen-iden3/src/helpers/identity-helpers.ts [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+  let arr
+  if (hexString) {
+    arr = Hex.decodeString(hexString)
+  } else {
+    arr = new Uint8Array(32)
+    window.crypto.getRandomValues(arr)
```
Comment (napalmpapalam) [inline packages/identity-gen-iden3/src/helpers/identity-helpers.ts:9]:
```ts
const generatePrivateKey = () => {
  const privateKey = new Uint8Array(32)

   if (!(typeof window === 'undefined')) {
     return window.crypto.getRandomValues(arr)
   }

   const webcrypto = require('crypto').webcrypto;
   return webcrypto.getRandomValues(privateKey);
}

export const getPrivateKey = (hexString?: string): string => {
  if (hexString) {
    return new PrivateKey(Hex.decodeString(hexString)).hex()
  }

    const privateKey = generatePrivateKey()
    return initPrivateKey(new PrivateKey(privateKey).hex())
}
```

File: packages/identity-gen-iden3/src/identity.ts [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+  CLAIM_PROOF_SIBLINGS_COUNT: number
+}
+
+export class Identity {
+  privateKeyHex = '' as string
+  identityId: Id = {} as Id
+  authClaimIncProofSiblings: Siblings = [] as Siblings
+  authClaimNonRevProofSiblings: Siblings = [] as Siblings
+  treeState: TreeState = {} as TreeState
+  coreAuthClaim: Claim = {} as Claim
```
Comment (napalmpapalam) [inline packages/identity-gen-iden3/src/identity.ts:40]: private?

File: packages/identity-gen-iden3/src/identity.ts [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+
+  public get privateKey() {
+    return new PrivateKey(Hex.decodeString(this.privateKeyHex))
+  }
+
+  public get identityIdString() {
+    return this.identityId.string()
+  }
+
+  public get identityIdBigIntString() {
```
Comment (napalmpapalam) [inline packages/identity-gen-iden3/src/identity.ts:72]:
```ts
 public get idString() {
    return this.identityId.string()
  }

  public get idBigIntString() {
```

File: packages/identity-gen-iden3/src/identity.ts [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+  }
+
+  public get authClaimInput() {
+    return [
+      ...this.coreAuthClaim.index.map(el => el.toBigInt().toString()),
+      ...this.coreAuthClaim.value.map(el => el.toBigInt().toString()),
+    ]
+  }
+
+  async createIdentity() {
```
Comment (napalmpapalam) [inline packages/identity-gen-iden3/src/identity.ts:83]: create method should return something

File: packages/identity-gen-iden3/src/identity.ts [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+      this.coreAuthClaim.getRevocationNonce(),
+      revocationsTreeRoot,
+    )
+
+    const authClaimNonRevProofSiblings = circomSiblingsFromSiblings(
+      authClaimNonRevProof.proof.siblings,
+      Identity.config.CLAIM_PROOF_SIBLINGS_COUNT,
+    )
+
+    // authClaimIncProof.proof.siblings = authClaimIncProofSiblings
```
Comment (napalmpapalam) [inline packages/identity-gen-iden3/src/identity.ts:137]: /

File: packages/identity-gen-iden3/src/identity.ts [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+
+    this.treeState = {
+      state: stateHash.string(),
+      claimsRoot: claimsTreeRoot.string(),
+      revocationRoot: revocationsTreeRoot.string(),
+      rootOfRoots: rootsTreeRoot.string(),
+    }
+  }
+
+  createCoreAuthClaim() {
```
Comment (napalmpapalam) [inline packages/identity-gen-iden3/src/identity.ts:156]: return type

File: packages/shared/src/api/api.ts [hunk 13]
```diff
@@ -10,7 +10,7 @@ export const CONFIG = {
   MAX_PAGE_LIMIT: 100,
 }
 
-const createApi = (baseUrl: string): JsonApiClient => {
+export const createApi = (baseUrl: string): JsonApiClient => {
```
Comment (napalmpapalam) [inline packages/shared/src/api/api.ts:13]: where do u use this

File: packages/zkp-gen-iden3/README.md [hunk 14]
```diff
@@ -0,0 +1,107 @@
+# @rarimo/zkp-gen-iden3
+These packages aim to provide developers with a set of commonly used functions and features for building web applications, such as handling big numbers, date manipulation, subscribing to and receiving notifications when certain events occur with EventEmitter, and more.
```
Comment (napalmpapalam) [inline packages/zkp-gen-iden3/README.md:2]: \

File: packages/zkp-gen-iden3/README.md [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+    requestId: '1',
+    /**
+     * Identity generated by @rarimo/identity-gen-iden3
+     */
+    identity: identityStoreSnap.identity as Identity,
+    /**
+     * Verifiable credentials generated by @rarimo/auth-zkp-iden3
+     */
+    verifiableCredentials:
+      verifiableCredentials as VerifiableCredentials<QueryVariableName>,
```
Comment (napalmpapalam) [inline packages/zkp-gen-iden3/README.md:52]: why use as here

File: packages/zkp-gen-iden3/src/enums/operators.ts [hunk 16]
```diff
@@ -0,0 +1,9 @@
+export enum ZKP_OPERATORS {
+  SKIP = '0',
+  EQUALS = '1',
+  LESS_THAN = '2',
+  GREATER_THAN = '3',
+  IN = '4',
+  NOTIN = '5',
+  NOT_EQUALS = '6',
+}
```
Comment (napalmpapalam) [inline packages/zkp-gen-iden3/src/enums/operators.ts:9]: pascal case

File: packages/zkp-gen-iden3/src/helpers/binary-helpers.ts [hunk 17]
```diff
@@ -0,0 +1,26 @@
+import { Constants, ElemBytes } from '@iden3/js-iden3-core'
+
+export function unmarshalBinary(data: Uint8Array): string[] {
+  const wantLen = 2 * Constants.ELEM_BYTES_LENGTH * Constants.BYTES_LENGTH
```
Comment (napalmpapalam) [inline packages/zkp-gen-iden3/src/helpers/binary-helpers.ts:4]: requiredLength

File: packages/zkp-gen-iden3/src/helpers/binary-helpers.ts [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+  }
+  const index = []
+  const value = []
+  for (
+    let i = 0, j = Constants.ELEM_BYTES_LENGTH;
+    i < Constants.ELEM_BYTES_LENGTH;
+    i++, j++
+  ) {
+    index[i] = new ElemBytes(
+      data.slice(i * Constants.BYTES_LENGTH, (i + 1) * Constants.BYTES_LENGTH),
```
Comment (napalmpapalam) [inline packages/zkp-gen-iden3/src/helpers/binary-helpers.ts:16]: extract

File: packages/zkp-gen-iden3/src/helpers/binary-helpers.ts [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+    )
+    value[i] = new ElemBytes(
+      data.slice(j * Constants.BYTES_LENGTH, (j + 1) * Constants.BYTES_LENGTH),
+    )
+  }
+  return [
+    ...index.map(el => el.toBigInt().toString()),
```
Comment (napalmpapalam) [inline packages/zkp-gen-iden3/src/helpers/binary-helpers.ts:23]: extract map cb

File: packages/zkp-gen-iden3/src/helpers/file-helpers.ts [hunk 20]
```diff
@@ -0,0 +1,8 @@
+import { fetcher } from '@distributedlab/fetcher'
+
+export async function readBytesFile(path: string) {
```
Comment (napalmpapalam) [inline packages/zkp-gen-iden3/src/helpers/file-helpers.ts:3]: why its duplicated?

File: packages/zkp-gen-iden3/src/types/contracts/common.ts [hunk 21]
```diff
@@ -0,0 +1,44 @@
+/* Autogenerated file. Do not edit manually. */
```
Comment (napalmpapalam) [inline packages/zkp-gen-iden3/src/types/contracts/common.ts:1]: duplicated?

File: packages/zkp-gen-iden3/src/types/contracts/factories/index.ts [hunk 22]
```diff
@@ -0,0 +1,4 @@
+/* Autogenerated file. Do not edit manually. */
+/* tslint:disable */
+/* eslint-disable */
+export { StateV2__factory } from "./StateV2__factory";
```
Comment (napalmpapalam) [inline packages/zkp-gen-iden3/src/types/contracts/factories/index.ts:4]: duplicated?

File: packages/zkp-gen-iden3/src/zkp-gen.ts [hunk 23]
```diff
@@ -0,0 +1,386 @@
...
+export class ZkpGen<T extends QueryVariableNameAbstract> {
+  requestId = ''
+  identity: Identity = {} as Identity
+  query: ZkpGenQuery<T> = {} as ZkpGenQuery<T>
+  verifiableCredentials: VerifiableCredentials<T> =
+    {} as VerifiableCredentials<T>
+
+  challenge = ''
+
+  subjectProof: ZKProof = {} as ZKProof
```
Comment (napalmpapalam) [inline packages/zkp-gen-iden3/src/zkp-gen.ts:57]: private?

File: packages/zkp-gen-iden3/src/helpers/state-v2-helpers.ts [hunk 24]
```diff
@@ -0,0 +1,22 @@
+import { providers } from 'ethers'
+
+import { StateV2__factory } from '@/types'
+
+export const getGISTProof = async ({
```
Comment (napalmpapalam) [inline packages/zkp-gen-iden3/src/helpers/state-v2-helpers.ts:5]: duplicated

File: packages/zkp-gen-iden3/src/zkp-gen.ts [hunk 25]
```diff
@@ -0,0 +1,386 @@
...
+  circomSiblingsFromSiblings,
+  hashElems,
+  newHashFromHex,
+  Proof,
+} from '@iden3/js-merkletree'
+import type { VerifiableCredentials } from '@rarimo/auth-zkp-iden3'
+import { type Identity } from '@rarimo/identity-gen-iden3'
+import { Buffer } from 'buffer'
+import type { BigNumber } from 'ethers'
+import omit from 'lodash/omit'
```
Comment (napalmpapalam) [inline packages/zkp-gen-iden3/src/zkp-gen.ts:28]: u dont need lodash just for case where u omit 1 field

File: packages/zkp-gen-iden3/src/zkp-gen.ts [hunk 26]
```diff
@@ -0,0 +1,386 @@
...
+            .revocationNonce,
+        ),
+      ),
+    )
+
+    return { path, coreClaim, claimProof }
+  }
+
+  async #getSchemaHash(): Promise<SchemaHash> {
+    const { data } = await fetcher.get<Schema>(
```
Comment (napalmpapalam) [inline packages/zkp-gen-iden3/src/zkp-gen.ts:323]: update `@distributedlab` packages version in each package

**Review by napalmpapalam — CHANGES_REQUESTED** [review]

File: packages/auth-zkp-iden3/README.md [hunk 27]
```diff
@@ -0,0 +1,68 @@
...
+  { find: "@iden3/js-crypto", replacement: "../../node_modules/@iden3/js-crypto/dist/esm_esbuild/index.js" },
+  { find: "@iden3/js-jsonld-merklization", replacement: "../../node_modules/@iden3/js-jsonld-merklization/dist/esm_esbuild/index.js" }
+]
+```
+
+## Running the tests
+
+```
+yarn test
+```
```
[the remark this answers is not in the document]
Comment (lukachi) [reply packages/auth-zkp-iden3/README.md:64]: no(

**Review by napalmpapalam — APPROVED** [review]

**Pull request merged**